File: basic/search/1064.py
from typing import List


class Solution:

    def fixedPoint(self, arr: List[int]) -> int:
        '''寻找满足 arr[i] == i 的最小索引 i'''
        l, r = 0, len(arr) - 1
        while l < r:
            m = (l + r) // 2
            if arr[m] < m:
                l = m + 1
            elif arr[m] == m:
                r = m  #要找最小索引
            else:
                r = m - 1
        if arr[l] != l:
            return -1
        else:
            return l


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(',')]
    solution = Solution()
    print(solution.fixedPoint(arr))

File: basic/search/1064.cpp
#include <iostream>

class Solution {
 public:
  int binary_search(int arr[], int len) {
    int l = 0, r = len - 1;
    while (l < r) {
      int m = l + (r - l) / 2;
      if (arr[m] < m) {
        l = m + 1;
      } else if (arr[m] > m) {
        r = m - 1;
      } else {
        r = m;
      }
    }
    return (arr[l] == l) ? l : -1;
  }
};

int main() {
  int a[7] = {-1, 1, 2, 3, 5, 8, 9};

  Solution solution = Solution();
  int idx = solution.binary_search(a, 7);
  std::cout << "idx: " << idx << std::endl;
}